### ISSUE TITLE
flatlist rerenders all list items when data changes

### ISSUE THREAD

**Issue by cwtowns** [microsoft/react-native-windows#7602]
## Environment
-  `npx react-native --version`:  5.0.1-alpha.2
 - `npx react-native info`:
System:
    OS: Windows 10 10.0.19042
    CPU: (6) x64 Intel(R) Core(TM) i5-9600K CPU @ 3.70GHz
    Memory: 19.25 GB / 31.92 GB
  Binaries:
    Node: 14.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.11 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.17763.0, 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.8.31019.35 (Visual Studio Community 2019)
  Languages:
    Java: Not Found
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.1 => 17.0.1
    react-native: 0.64.0 => 0.64.0
    react-native-windows: ^0.64.0-0 => 0.64.3
  npmGlobalPackages:
    *react-native*: Not Found

- Target Device(s):  Desktop
- Visual Studio Version:  2019

## Steps To Reproduce
The following code recreates the problem:

```
import React, {useState, useEffect, useMemo} from 'react';
import {SafeAreaView, Text, FlatList, ListRenderItem} from 'react-native';

interface LogMessage {
  message: string;
}

const LogElement = (props: LogMessage) => {
  return useMemo(() => {
    return <Text>{props.message}</Text>;
  }, [props.message]);
};

const FlatListRendersTooMuch = () => {
  const [flatListContents, setFlatListContents] = useState([] as LogMessage[]);

  useEffect(() => {
    console.log('mounting');
    const id = setInterval(() => {
      console.log('adding new message');
      setFlatListContents(f =>
        f.concat({
          message: new Date().toString(),
        }),
      );
    }, 5000);
    return () => {
      console.log('unmounting');
      clearInterval(id);
    };
  }, []);

  const keyExtractor = (item: LogMessage) => {
    return item.message;
  };

  const renderItem: ListRenderItem<LogMessage> = ({item}) => {
    console.log('renderItemMemoFucntion ' + item.message);
    return <LogElement {...item} />;
  };

  const renderItemMemoized = useMemo(() => renderItem, []);

  return useMemo(() => {
    return (
      <SafeAreaView>
        <FlatList
          style={{height: 150, paddingTop: 10}}
          data={flatListContents}
          renderItem={renderItemMemoized}
          keyExtractor={keyExtractor}
        />
      </SafeAreaView>
    );
  }, [flatListContents, renderItemMemoized]);
};

export default FlatListRendersTooMuch;
```

I believe this follows best practices for optimal FlatList performance by:

- providing a key extractor
- memoizing the render item JSX function
- memoizing the render item function itself

However the list renders more than expected (as shown by the log statements in the renderItem method). For example, here is the output when the first element is added to the list:

```
adding new message 
renderItemMemoFucntion Wed Apr 14 2021 16:56:13 GMT-0500 (Central Daylight Time)
```

This is the output when the second message is added:
```
adding new message
renderItemMemoFucntion Wed Apr 14 2021 16:56:13 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:18 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:13 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:18 GMT-0500 (Central Daylight Time)
```

And this is the output when the third message is added:
```
adding new message
renderItemMemoFucntion Wed Apr 14 2021 16:56:13 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:18 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:23 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:13 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:18 GMT-0500 (Central Daylight Time)
renderItemMemoFucntion Wed Apr 14 2021 16:56:23 GMT-0500 (Central Daylight Time)
```

## Expected Results

Adding a new item should only show render calls being made for that new item.  Previously added items should not be rendered again.  

## Snack, code example, screenshot, or link to a repository:
https://snack.expo.io/2zp0tqSKR

**Comment by rozele** [comment]
This is already "fixed" by #7470. Perhaps it's worth cherry-picking back to 0.64. Thoughts @chrisglein?

I say "fixed" because there is also a potential optimization in Yoga to modify the layout algorithm so it doesn't recalculate layout for stretch aligned children when the cross-axis dimension of the parent has not changed, but I'm still investigating that one 😅.

**Comment by rectified95** [comment]
Fix cherry-picked into 0.64-stable.
